src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -0,0 +1,256 @@
+  Transaction& t,
+  ool_record_t& record) {
+  bufferlist bl = record.encode(current_segment->segment->get_segment_id(), 0);
+  seastar::promise<> pr;
+  current_segment->inflight_writes.emplace_back(pr.get_future());
+
+  logger().debug(
+    "SegmentedAllocator::Writer::write: written {} extents,"
+    " {} bytes to segment {} at {}",
+    record.get_num_extents(),
+    bl.length(),
+    current_segment->segment->get_segment_id(),
+    record.get_base());
+
+  return trans_intr::make_interruptible(
+    current_segment->segment->write(record.get_base(), bl).safe_then(
+      [this, pr=std::move(pr),
+      it=(--current_segment->inflight_writes.end()),
+      cs=current_segment]() mutable {
+        if (cs->outdated) {
+          pr.set_value();
+        } else{
+          current_segment->inflight_writes.erase(it);
+        }
+        return seastar::now();
+    })
+  ).si_then([this, &record, &t]() mutable {
+    return finish_write(t, record);
+  });
+}
+
+void SegmentedAllocator::Writer::add_extent_to_write(
+  ool_record_t& record,
+  LogicalCachedExtentRef& extent) {
+  extent->prepare_write();
+  // all io_wait_promise set here will be completed when the corresponding
+  // transaction "complete_commit"s, we can't do complete_io() here, otherwise,
+  // there might be consistency problem.
+  record.add_extent(extent);
+}
+
+SegmentedAllocator::Writer::write_iertr::future<>
+SegmentedAllocator::Writer::write(
+  Transaction& t,
+  std::list<LogicalCachedExtentRef>& extents)
+{
+  auto fut = trans_intr::make_interruptible(roll_segment_ertr::now());
+  if (rolling_segment) {
+    fut = trans_intr::make_interruptible(
+      (roll_segment_ertr::future<>)rolling_segment->get_shared_future());
+  } else if (!current_segment) {
+    fut = roll_segment(true);
+  }
+  return fut.si_then([this, &extents, &t] {
+    return seastar::do_with(ool_record_t(segment_manager.get_block_size()),
+      [this, &extents, &t](auto& record) {
+      return trans_intr::repeat([this, &record, &t, &extents] {
+        if (extents.empty()) {
+          return (write_iertr::future<seastar::stop_iteration>)
+            seastar::make_ready_future<
+              seastar::stop_iteration>(seastar::stop_iteration::yes);
+        }
+        auto fut = trans_intr::make_interruptible(roll_segment_ertr::now());
+        if (rolling_segment) {
+          fut = trans_intr::make_interruptible(
+            (roll_segment_ertr::future<>)rolling_segment->get_shared_future());
+        }
+        return fut.si_then([this, &record, &extents, &t] {
+          record.set_base(allocated_to);
+          for (auto it = extents.begin();
+               it != extents.end();) {
+            auto& extent = *it;
+            auto wouldbe_length =
+              record.get_wouldbe_encoded_record_length(extent);
+            if (_needs_roll(wouldbe_length)) {
+              // reached the segment end, write and roll
+              assert(!rolling_segment);
```

Comment:
IIUC `rolling_segment->get_shared_future()` returns a shared future that might have multiple waiters.

So, is it possible that the first ready waiter enters `_needs_roll()`, set `rolling_segment`, then `_write()`. Then a second ready waiter also enters `_needs_roll()` and got assertion failure?